dscribe/ext/soapGTO.cpp

```diff
@@ -2059,152 +2062,176 @@ void getCD(
-              C_mu( posI, typeJ,n,m) += bOa[LNsNs + n*Ns + k]*sumMe;
-             }
+      shift = 0;
+      for (int k = 0; k < Ns; k++) {
+        for (int m = restOfLs*restOfLs; m < (restOfLs+1)*(restOfLs+1); m++) {
+          sumMe = 0;
+          for (int i = 0; i < Asize; i++) {
+            preExp = preExponentArray[Asize*k+i];
+            sumMe += preExp*preCoef[totalAN*(m-4)+i];
+            shift++;
           }
+          for (int n = 0; n < Ns; n++) {
+            C_mu(posI, typeJ, n, m) += bOa[LNsNs + n*Ns + k]*sumMe;
+          }
+        }
       }
 
-           if(return_derivatives){
-             shift = 0;
-      for(int k = 0; k < Ns; k++){
-        for(int i = 0; i < Asize; i++){
-          preExp = preExponentArrya[shift];
+      if (return_derivatives) {
+        shift = 0;
+        for (int k = 0; k < Ns; k++) {
+          for (int i = 0; i < Asize; i++) {
+            preExp = preExponentArray[shift];
             shift++;
-          for(int m = restOfLs*restOfLs; m < (restOfLs+1)*(restOfLs+1); m++){
-            preVal = 2.0*aOa[LNs + k]*preExp*preCoef[totalAN*(m-4)+i];
-            preValX = x[i]*preVal + preExp*prCofDX[totalAN*(m-4)+i];
-            preValY = y[i]*preVal + preExp*prCofDY[totalAN*(m-4)+i];
-            preValZ = z[i]*preVal + preExp*prCofDZ[totalAN*(m-4)+i];
-            for(int n = 0; n < Ns; n++){
-          CDevX_mu(indices[i], posI, typeJ,n,m) += bOa[LNsNs + n*Ns + k]*preValX;
-          CDevY_mu(indices[i], posI, typeJ,n,m) += bOa[LNsNs + n*Ns + k]*preValY;
-          CDevZ_mu(indices[i], posI, typeJ,n,m) += bOa[LNsNs + n*Ns + k]*preValZ;
-          
-               }
+            for (int m = restOfLs*restOfLs; m < (restOfLs+1)*(restOfLs+1); m++) {
+              preVal = 2.0*aOa[LNs + k]*preExp*preCoef[totalAN*(m-4)+i];
+              preValX = x[i]*preVal + preExp*prCofDX[totalAN*(m-4)+i];
+              preValY = y[i]*preVal + preExp*prCofDY[totalAN*(m-4)+i];
+              preValZ = z[i]*preVal + preExp*prCofDZ[totalAN*(m-4)+i];
+              for (int n = 0; n < Ns; n++) {
+                CDevX_mu(indices[i], posI, typeJ, n, m) += bOa[LNsNs + n*Ns + k]*preValX;
+                CDevY_mu(indices[i], posI, typeJ, n, m) += bOa[LNsNs + n*Ns + k]*preValY;
+                CDevZ_mu(indices[i], posI, typeJ, n, m) += bOa[LNsNs + n*Ns + k]*preValZ;
+              }
             }
+          }
         }
       }
     }
   }
-}
-free(preExponentArrya);
+  free(preExponentArray);
+
+  // If attach=True, the derivative with respect to the center atom coordinates
+  // is the negative sum of derivatives with respect to coordinates of other
+  // atoms in the the neighbourhood.
+  if (return_derivatives && attach && (posAtomI >= 0)) {
+    for (int m = 0; m < (lMax+1)*(lMax+1); m++) {
+      for (int n = 0; n < Ns; n++) {
+        double sumX = 0;
+        double sumY = 0;
+        double sumZ = 0;
+        for (int i = 0; i < Asize; i++) {
+          if (indices[i] != posAtomI) {
```

Comment:
WIll there be a problem if the user has excluded some atoms from the derivative calculation using either the `include` or `exclude` argument? Say one of the neighbouring atoms of the center of interest has been excluded: the derivatives for it will not be calculated (=they are zero).

If this is a problem, then the easiest fix might be that you raise an exception if the there is a combination of `attach=True` and `indices` does not contain all atoms. A better solution may be to add all the neighbours of included atoms to the derivative calculation, but not report their derivative values in the final result.

Reply:
I have tested with `include`/`exclude` and it still works. The reason is that the current implementation (even when `attach=False`) computes the derivatives of coefficients c with respect to the coordinates of all atoms in the center's neighbourhood. In fact, the list of included atoms (`indices_u`) is not even used in function `soapGTO()` before calling `getPDev()`, at which point the c-derivatives have already been computed.

I am quite sure that this is not optimal, and in the case of `attach=False`, it would be enough to differentiate c with respect to atoms in the neighbourhood that are included in `indices_u`. If this optimization is ever made, the `attach=True` approach has to be revised. In that case I think that the second option that you mention would be the way to go.